apps/system/js/acl_manager.js

```diff
@@ -0,0 +1,283 @@
+    for (var size in app.manifest.icons) {
+      if (!smallestSize || size < smallestSize) {
+        smallestSize = size;
+      }
+    }
+    var iconURL = app.manifest.icons[smallestSize];
+    if (iconURL.startsWith('/')) {
+      iconURL = app.origin + iconURL;
+    }
+    return iconURL;
+  };
+
+  ACLManager.prototype.handleMessage = function(msg) {
+    debug('processing message', msg);
+
+    try {
+      msg = JSON.parse(msg);
+    } catch (e) {
+      console.log('ACL: Unable to deserialize message');
+      return;
+    }
+
+    var app = this.getAppFromMessage(msg);
+    if (!app) {
+      console.log('ACL: Unable to fetch app from message');
+      return;
+    }
+
+    switch (msg.action) {
+      case 'launch':
+        this.launchApp(app);
+        break;
+
+      case 'kill':
+        this.killApp(app);
+        break;
+
+      case 'minimize':
+        this.minimizeApp(app);
+        break;
+
+      case 'notify':
+        this.sendNotification(app, msg);
+        break;
+
+      case 'notify-remove':
+        this.removeNotification(msg.id);
+        break;
+
+       default:
+        debug('unrecognized action', msg.action);
+        break;
+    }
+  };
+
+  ACLManager.prototype.launchApp = function(app) {
+    window.dispatchEvent(new CustomEvent('webapps-launch', {
+      detail: {
+        manifestURL: app.manifestURL,
+        url: app.origin + app.manifest.launch_path,
+        timestamp: Date.now()
+      }
+    }));
+  };
+
+  ACLManager.prototype.killApp = function(app) {
+    if (this.isExternalApp(app.manifest)) {
+      AppWindowManager.kill(app.origin, app.manifestURL);
+    }
+  };
+
+  ACLManager.prototype.minimizeApp = function(app) {
+    // Only minimize requested app if it is active.
+    if (AppWindowManager.getActiveApp().manifestURL === app.manifestURL) {
+      window.dispatchEvent(new CustomEvent('home'));
+    }
+  };
+
+  ACLManager.prototype.sendNotification = function(app, detail) {
+    if (!detail.id) {
+      debug('No id received with notification, cannot fire click callback');
+      detail.id = getRandomString() + Date.now();
+    }
+
+    if (!detail.icon) {
+      detail.icon = this.getAppIcon(app);
+    }
+
+    if (detail.silent) {
+      detail.icon += '#silent=1';
+    }
+
+    var n = new Notification(detail.title || '', {
+      body: detail.body || '',
+      icon: detail.icon,
+      tag: NOTIFICATION_PREFIX + detail.id
+    });
+
+    n.onclick = function() {
+      this.sendMessage({
+        action: 'notify-click',
+        id: detail.id,
+        origin: detail.origin
+      });
+    }.bind(this);
+
+    this.notifications[detail.id] = n;
+  };
+
+  ACLManager.prototype.removeNotification = function(id) {
+    if (!id || !this.notifications[id]) {
+      debug('Unable to find notification for removal', id);
+    } else {
+      this.notifications[id].close();
+      delete this.notifications[id];
+    }
+  };
+
+  exports.aclManager = new ACLManager();
```

Comment:
nit: I think that these two lines should be in bootstrap.js? Generally they're placed there to help with unit testing and such.